sky/engine/core/painting/painting.cc

```diff
@@ -0,0 +1,89 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "sky/engine/core/painting/painting.h"
+
+#include "base/bind.h"
+#include "base/logging.h"
+#include "base/trace_event/trace_event.h"
+#include "sky/engine/core/painting/CanvasImage.h"
+#include "sky/engine/platform/SharedBuffer.h"
+#include "sky/engine/platform/image-decoders/ImageDecoder.h"
+#include "sky/engine/platform/mojo/data_pipe.h"
+#include "sky/engine/tonic/dart_invoke.h"
+#include "sky/engine/tonic/dart_persistent_value.h"
+#include "sky/engine/tonic/mojo_converter.h"
+
+namespace blink {
+namespace {
+
+void DecodeImage(scoped_ptr<DartPersistentValue> callback,
+		 PassRefPtr<SharedBuffer> buffer) {
+  TRACE_EVENT0("blink", "DecodeImage");
+
+  DartState* dart_state = callback->dart_state().get();
+  if (!dart_state)
+    return;
+  DartState::Scope scope(dart_state);
+
+  OwnPtr<ImageDecoder> decoder =
+      ImageDecoder::create(*buffer.get(), ImageSource::AlphaPremultiplied,
+                           ImageSource::GammaAndColorProfileIgnored);
+
+  // decoder can be null if the buffer was empty and we couldn't even guess
+  // what type of image to decode.
+  if (!decoder) {
+    DartInvokeAppClosure(callback->value(), {nullptr});
+    return;
+  }
+  decoder->setData(buffer.get(), true);
+  if (decoder->failed() || decoder->frameCount() == 0) {
+    DartInvokeAppClosure(callback->value(), {nullptr});
+    return;
+  }
+
+  RefPtr<CanvasImage> resultImage = CanvasImage::create();
+  ImageFrame* imageFrame = decoder->frameBufferAtIndex(0);
+  RefPtr<SkImage> skImage = adoptRef(
+      SkImage::NewFromBitmap(imageFrame->getSkBitmap()));
+  resultImage->setImage(skImage.release());
+
+  Dart_Handle resultHandle = DartConverter<CanvasImage*>::ToDart(
+      resultImage.get());
```

Comment:
You can just write `ToDart(resultImage.get())`.  There's template magic to automatically detect the type.